gossip/src/cluster_info_entry_listener.rs

```diff
@@ -0,0 +1,133 @@
+use {
+    crate::{
+        cluster_info::{ClusterInfo, GOSSIP_SLEEP_MILLIS},
+        crds::Cursor,
+        crds_value::CrdsData,
+    },
+    log::*,
+    std::{
+        sync::{
+            atomic::{AtomicBool, Ordering},
+            Arc,
+        },
+        thread::{self, sleep, Builder, JoinHandle},
+        time::Duration,
+    },
+};
+
+pub trait ClusterInfoEntryHandler: Send {
+    fn handle(&mut self, data: CrdsData);
+}
+
+pub struct ClusterInfoEntryListener {
+    thread_hdls: Vec<JoinHandle<()>>,
+}
+
+// right now we only need to process duplicate proof, in the future the receiver
+// should be a map from enum value to handlers.
+impl ClusterInfoEntryListener {
+    pub fn new(
+        exit: Arc<AtomicBool>,
+        cluster_info: Arc<ClusterInfo>,
+        duplicate_proof_handler: impl ClusterInfoEntryHandler + 'static,
+    ) -> Self {
+        let listen_thread = {
+            Builder::new()
+                .name("solCiEntryLstnr".to_string())
+                .spawn(move || {
+                    Self::recv_loop(exit, &cluster_info, duplicate_proof_handler);
+                })
+                .unwrap()
+        };
+
+        Self {
+            thread_hdls: vec![listen_thread],
+        }
+    }
+
+    pub fn join(self) -> thread::Result<()> {
+        self.thread_hdls.into_iter().try_for_each(JoinHandle::join)
+    }
+
+    // Here we are sending data one by one rather than in a batch because in the future
+    // we may send different type of CrdsData to different senders.
+    fn recv_loop(
+        exit: Arc<AtomicBool>,
+        cluster_info: &ClusterInfo,
+        mut duplicate_proof_handler: impl ClusterInfoEntryHandler + 'static,
+    ) {
+        let mut cursor = Cursor::default();
+        while !exit.load(Ordering::Relaxed) {
+            let entries: Vec<CrdsData> = cluster_info.get_duplicate_shred_entries(&mut cursor);
+            inc_new_counter_debug!("cluster_info_entry_listener-recv_count", entries.len());
+            for x in entries {
+                duplicate_proof_handler.handle(x);
+            }
```

Comment:
Agree with Carl here.
Not sure we need `cluster_info_entry_listener.rs` at the moment.
I believe the crds `Cursor` and the inner indices already provide this functionality.
Similar to how these 3 functions are implemented:
https://github.com/solana-labs/solana/blob/d1722b7a7/gossip/src/crds.rs#L317-L353

Reply:
The point of adding cluster_info_entry_listener is just that we don't have to keep adding new index, new buckets, and new cluster_info_xxx_listener to our code every time we have a new type of events to ingest from Gossip. Compare the current plan when we are ingesting a new type of foo event:
1. add a new foo event index in crds and all related processing
2. add a cluster_info_foo_listener so it has its own thread to call cluster_info.get_foo() to get all foo events between GOSSIP_SLEEP_MILLIS using the cursor
3. add foo handling logic in cluster_info_foo_listener
with the new plan after adding cluster_info_event_listener
1. add your own handling logic in foo_handler
2. pass the handler in the cluster_info_event_listener (this would be one line change on the command line)

The new plan doesn't touch crds and everything below cluster_info_event_listener at all, the event dispatching is done solely in cluster_info_event_listener, and the foo specific logic is clearly separated in handler vs the dispatch logic in cluster_info_event_listener.

That said, I do realize how much we hate Gossip and how we are very against adding even more event types into Gossip, so I'm okay with sticking to the current plan and make one more index in crds and create cluster_info_duplicate_entry_listener.rs accordingly, I'll change the code to do that.

Reply:
A single generic entry_listener would require all those handlers to run sequentially within the same thread; so unrelated handlers can potentially block each other which is not desired.

Reply:
The number of threads used is orthogonal to the receiver->dispatcher->handler stack. You could of course start multiple threads in the dispatcher, but this is normally too intrusive because it requires the dispatcher to understand how many threads each handler would need and how costly the packets are, this is application level knowledge which shouldn't pollute the underlying stack. Normally the dispatcher would do minimum work and send the data in some zero-copy way to the handlers, the handlers could use as many threads as they fit.

Admitted, receiver->dispatcher->handler stack could be overkill in some cases, but it does offer a nice function separation.

Reply:
If the computations for handling different types of entries (i.e. votes vs epoch-slots vs duplicates shreds, etc) don't want to block each other, then each should run on a different thread.
But if they are already running on a different thread then they can just directly utilize the existing and more efficient crds index+Cursor machinery; a generic listener would only add more overhead there.

It really does not take much to add a new index for a certain type, as done in https://github.com/solana-labs/solana/pull/29317 for duplicate shreds.

Reply:
No worries, as I said, receiver->dispatcher->handler stack is better if you want to keep adding new packets/event types to this stack, as it's more extensible. But I do understand we don't want to add more types to Gossip, so I'm okay with sticking to the current flow. I'm not against the method you proposed, I just haven't got to this change yet.

Reply:
Changed